### PR TITLE
1068 Generating context docs from schema files

### LINKED ISSUE #1068: Replace existing Context pages with JSON schema definitions
### Use Case:
Generate Docusaurus pages for Context types using JSON schema files.

Proposal already created for markdown docs, which can be found here #1020 , which is included as part of release 2.1

### Work required
Either:
- Script to generate static markdown documentation from JSON schema files
- Build or integrate a JSON Schema viewer solution

and:
- Generate the Docusaurus front matter, programmatically modify the sidebars file to make sure that they're linked and they're in the website or the viewer.

### Additional Information
Discussion around this particular enhancement is being tracked via #940 and #1059

### PR DESCRIPTION
**Pull request by TheJuanAndOnly99**
resolves #1068 

First iteration is complete:
- Created a script that generates markdown files from schema files; to execute run `yarn schema2markdown` in the root directory.
- The script generates context and api markdown docs from schemas
- Links are generated for each reference encountered
- Sidebars entries are automatically appended to sidebars.json
- The script iterates across all versions
- Each object definition maps to a file, as opposed to the previous version

Todo:
- confirm markdown page layout
- code cleanup

Preview - https://issue-1068--exquisite-otter-12d363.netlify.app/docs/context/schemas/Action

### CODE REVIEW

**Comment by TheJuanAndOnly99** [comment]
@kriswest @robmoffat - first iteration is done; please check the description above to see what's been done and what not; there's also a preview to a netlify pointing to my fork (the build command will need to be updated when we go live with this).

Eager to hear your feedback! /cc @mistryvinay

File: schema2Markdown.js
Comment (julianna-ciq) [inline schema2Markdown.js:223]: Is it necessary to hardcode the version numbers? Do they live somewhere else that we could get them from, such as in the directory structure? Just wondering, for future-proofing.

File: schema2Markdown.js
Comment (julianna-ciq) [inline schema2Markdown.js:227]: Personally, I'd prefer not using `require`, just because it's caused me problems when I've managed system changes/upgrades in the past. I don't have anything specific to point t, though.

Instead, you could use fs-extra's `readJsonSync`, to match the later use of `outputJsonSync`

File: schema2Markdown.js
Comment (julianna-ciq) [inline schema2Markdown.js:1]: If you're going to use `fs-extra`, could you name it `fse`, or something else to differentiate it from the Node `fs` package? I kept reading things like `fs.outputJsonSync` and thinking, "when did fs add *that* method?".

**Review by julianna-ciq — CHANGES_REQUESTED** [review]

**Comment by maoo** [comment]
Hi @julianna-ciq , thanks SO much for the detailed review! @TheJuanAndOnly99 is currently off, he'll be back on March 3rd and work on your comments.

File: schema2Markdown.js
Comment (julianna-ciq) [inline schema2Markdown.js:11]: I think you could use flow control to remove the `else` block (reducing a nesting layer), if you'd like. I think something like this would be easier to understand/maintain, but it's a personal preference.

(Not saying I'm opposed to `else` blocks. I just think *this* one in particular is unnecessary.)

```js
if (propertyName === 'type') {
    markdownContent += `### Type\n\n`;
    markdownContent += `\`${propertyDetails.const}\`\n\n`;
    return markdownContent;
}

markdownContent += `### ${propertyDetails.title || propertyName}\n\n`;
markdownContent += `\`${propertyName}\`\n\n`;
// rest of the function
```

File: schema2Markdown.js
Comment (julianna-ciq) [inline schema2Markdown.js:62]:
```suggestion
    const objectType = filePath.split('/').pop().split('.')[0]; // api, AppIdentifier
```

File: schema2Markdown.js
Comment (julianna-ciq) [inline schema2Markdown.js:171]:
```suggestion
    return text.replace(/>/g, '\\>');
```

Personal preference. I think this would tighten the function up a bit without compromising readability. Also, the `replace` method doesn't mutate the original string [[citation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#description)], so we don't have to worry about side effects.

**Review by julianna-ciq — APPROVED** [review]
I have a couple of stylistic nitpicks, but I think it looks good. I think @kriswest or @robmoffat have to review as well, though.

**Review by robmoffat — APPROVED** [review]
Could we have a link to the schema file on the page, so that the user can view that?

Also, might be good to reference the Typescript type that is generated, if that's possible (and c# if that's possible too I guess)

**Comment by kriswest** [comment]
> Can you give me a clear mapping between the schema and the typescript (and c#) files to point to?

@TheJuanAndOnly99  The mapping from schema to typescript is easy because this script: https://github.com/finos/FDC3/blob/3c2af8b33c203f46a2b4cd6e5de48fa558a78538/package.json#L28
will generate this file: https://github.com/finos/FDC3/blob/main/src/context/ContextTypes.ts

which contains typescript for every schema in the /schemas/context folder. 

However, that doesn't give you a simple link to use from a particular docs page to the TS, but it does mean we can add some advice somewhere about where to get TypeScript types from - which I am surprised to report that we are not actually doing anywhere! It should probably be added as a section in the context overview:
https://fdc3.finos.org/docs/context/spec

**Comment by robmoffat** [comment]
makes sense to me @maoo

File: schema2Markdown.js
Comment (kriswest) [inline schema2Markdown.js:181]: No need to add 'Schema'
```suggestion
function generateFrontMatter(title, description) {
    return {
        title: title,
        description: description,
        sidebar_label: title,
    };
}
```

File: schema2Markdown.js
Comment (kriswest) [inline schema2Markdown.js:227]:
```suggestion
            "label": "Context Data Part",
```

File: schema2Markdown.js
Comment (kriswest) [inline schema2Markdown.js:237]: We don't need a reference generating for this yet - when we do it'll probably be incorporated into the FDC3 protocol spec and handled slightly differently

```suggestion
```

File: schema2Markdown.js
Comment (kriswest) [inline schema2Markdown.js:12]: As its so important, we should move the type property up out of properties. Hence, just skip it here:

```suggestion
    if (propertyName === 'type') {
        //skip rendering the type property as it should be rendered at the top level
        return markdownContent;
    }   
```

File: schema2Markdown.js
Comment (kriswest) [inline schema2Markdown.js:132]: Redner type at top level and skip under properties section:

```suggestion
        const properties = root.properties;
        const typeString = properties?.type?.const;
        const required = root.required;
        const ref = root.$ref;
                
        markdownContent += `## Type\n\n`;
        markdownContent += `\`${typeString}\`\n\n`;

        markdownContent += `## Properties\n\n`; 

        for (const [propertyName, propertyDetails] of Object.entries(properties)) {
            if (propertyName != "type"){ 
                markdownContent += processProperty(propertyName, propertyDetails, schema.examples);
            }
        }
```

**Review by kriswest — CHANGES_REQUESTED** [review]
@TheJuanAndOnly99 Apologies for the delay in getting you a review here. Lots of good stuff, but we need a few tweaks before merging. I've dealt with as many as I can in comments/suggestions, but there are a few extras to handle:

- There is an issue with the rendering of the base [Context schema](https://issue-1068--exquisite-otter-12d363.netlify.app/docs/context/schemas/Context), which is a special case as all the others derive from it. Note that only definitions/DocumentedContext should be rendered for that file.

- The PR should go ahead and remove the existing ref pages and replace them with the generated content - we should check that it is generating the same URLs as the previous manually defined pages to avoid any broken links.

FInally, this PR could move to the FDC3 repo (rather than your fork) for final tweaks and merging. Could you apply suggestions, check they are ok and let me know? I'll then merge it into an FDC3 repo branch and get the other maintainers to sign off on merging.

**Comment by kriswest** [comment]
@TheJuanAndOnly99 merging changes from upstream:master/main and re-running npm install will probably resolve the issues with the checks you're having above. This should sort out your fork:

```
git fetch
git branch --move master main
git push -u origin main
```

Then update main from upstream:main and then merge into your branch.

**Comment by kriswest** [comment]
@TheJuanAndOnly99 one more issue we spotted in an important schema (instrument):
- New: https://issue-1068--exquisite-otter-12d363.netlify.app/docs/context/schemas/Instrument
- Old: https://fdc3.finos.org/docs/next/context/ref/Instrument

First, ignore the difference in the description - I moved the notes into the id property in the schema as they all related to that. However, what we've lost is the details of the standardized subproperties of `id` and of `market`. Its going to need to pick those up and show them in the generated page...

Hence, if a property is an object or array, you probably need to descend into it and handles its own properties, unless its a ref to another defined type...

**Comment by maoo** [comment]
> @TheJuanAndOnly99 merging changes from upstream:master/main and re-running npm install will probably resolve the issues with the checks you're having above. This should sort out your fork:
> 
> ```
> git fetch
> git branch --move master main
> git push -u origin main
> ```
> 
> Then update main from upstream:main and then merge into your branch.

@kriswest - we did that, but since then all my commits fail with `/Users/m/w/FDC3/src/internal/typeHelpers.ts -  Parsing error: Cannot read properties of undefined (reading 'map')` ; was there any change to this file recently? Any tips on how to address this?

**Comment by kriswest** [comment]
> @kriswest - we did that, but since then all my commits fail with `/Users/m/w/FDC3/src/internal/typeHelpers.ts - Parsing error: Cannot read properties of undefined (reading 'map')` ; was there any change to this file recently? Any tips on how to address this?

Yes, a PR got merged today that added that.  I'll see if I can replicate quickly.

**Comment by kriswest** [comment]
I can replicate. Its from the linter, which is setup to lint src on commit, where the lint command only does src/api. Not exactly sure what its problem is however...

**Comment by kriswest** [comment]
@maoo see https://github.com/finos/FDC3/pull/1187 for a solution (needs review and merge)

**Comment by kriswest** [comment]
...and you might need to delete package-lock.json

**Comment by kriswest** [comment]
Is this ready for re-review @maoo and @TheJuanAndOnly99 ? Its still marked draft so just checking. If its ready I can encourage the @finos/fdc3-maintainers and @finos/fdc3-editors to review and merge

File: schema2Markdown.js
Comment (kriswest) [inline schema2Markdown.js:218]: trips a semgrep rule as the first arg is passed to path.join a few lines below...  I can see that you don't actually take any user input so it either needs:
- a semgrep exception comment: https://semgrep.dev/docs/ignoring-files-folders-code
- don't take the folder as a parameter and just hardcode it to look at the context dirs (a comment at the top of the file that it only processes the context folders would be worth adding).

**Comment by kriswest** [comment]
@TheJuanAndOnly99 schema2markdown is tripping one of the semgrep rules: https://github.com/finos/FDC3/actions/runs/8837630351/job/24266920820?pr=1151

and the build failed, probably indicating that you need to run npm install in your branch (possibly also delete the local file first): https://github.com/finos/FDC3/actions/runs/8837630352/job/24266920775?pr=1151

**Comment by kriswest** [comment]
@TheJuanAndOnly99 you're all clear! Just he axios dependency issue in the log

**Comment by kriswest** [comment]
@mistryvinay and I should be able to take it from here - we need to run through another review and then make the tweaks necessary to have the generated pages replace the existing ones, meaning they'll need to generate onto the same URLs as the manual pages, which we'll delete!

**Comment by kriswest** [comment]
...although @TheJuanAndOnly99 you might need to explain or handle whatever is needed regarding the netlify preview/build... I don't yet know what change is needed nor why.

If its about making sure the generation script gets run, we could possibly do that with a commit hook and npm package script...

**Comment by kriswest** [comment]
@TheJuanAndOnly99 The preview for this appears to be broken - could you take a look, not seeing the generated pages anymore. Perhaps a conflict on updating from @robmoffat's homepage update

**Comment by TheJuanAndOnly99** [comment]
Hi @kriswest the preview is running on my personal Netlify (since the build command is different). You can find it [here](https://exquisite-otter-12d363.netlify.app/).

**Comment by kriswest** [comment]
@TheJuanAndOnly99 I'm not seeing the generated pages in the preview's sidebar anymore:
image

These are the old pages?, right: https://exquisite-otter-12d363.netlify.app/docs/next/context/ref/Action

**Comment by TheJuanAndOnly99** [comment]
@kriswest yep you are correct! I think I've fixed it.

**Review by hughtroeger — COMMENTED** [review]
Code looks fine. I went through the Context pages comparing the most recently provided preview to the current docs, here are my notes

General differences:
-	No Overview page for Context Data Part
-	No Context page for Context Data Part 
-	‘Required’ boolean field is missing for all properties
-	Schema link goes to github.com instead of fdc3.finos.org, assume that will update on publish/deploy
-	All reference links are broken in the Netlify preview, is this expected? 
-	All properties with array types no longer have specific array type (like Instrument[]), they just say ‘array’. Many do have a sentence that says what kind of array it is, and some have reference links to the schema page for the thing it is an array of
-	Properties with a reference link to another docs page no longer have ‘type’ explicitly declared, this is probably fine as the information is present in the page linked
-	I find the new format a little difficult to read, I think some indentation for the data under each property would go a long way.

Page specific differences:
-	Chart
o	Instruments property used to have type ‘Instrument[]’ now it has ‘array, is clear from the description that this is an array of instruments
-	ChatInitSettings
o	Message property used to be of type ‘Message’, new one has an example but should probably have a reference link to fdc3.message
-	Chat Search Criteria
o	Properties property used to have type 	(Instrument | Contact | Organization | string)[] , now it just has type ‘array’
-	Contact
o	No longer has ‘name’ property
-	ContactList
o	No longer has ‘id’ or ‘name’ properties
o	‘contacts’ property could have a reference link to the contact schema page
-	Country
o	No longer has ‘name’ property
-	Email
o	‘recipients’ property used to be of type ‘fdc3.contact or fdc3.contactList’, now it has no type and no reference link
-	Instrument
o	‘id’ subproperties used have working links under ‘More Info’, now they have URLs that aren’t active hyperlinks links displayed like https://www.isin.org/
-	InstrumentList
o	No longer has ‘name’ or ‘id’ properties
o	‘instruments’ property could have a reference link to instrument schema page
-	Order
o	Details.product has type and description ‘undefined’
-	OrderList
o	No longer has ‘name’ or ‘id’ properties
o	‘orders’ property used to be of type ‘Trade[]’ which I believe is incorrect, I think it should be Order[], now it just has type ‘array’ but does not have a reference link to the order schema page
-	Organization
o	Missing ‘name’ property
-	Portfolio
o	Missing ‘name and ‘id’ properties
o	‘positions’ property could link to position schema page
-	Position
o	Missing ‘name ‘ and ‘id’ properties
-	TradeList
o	Missing ‘id’ and ‘name’ properties
o	Trades property could link to the trades schema page
-	TransactionResult
o	Missing ‘message’ property

**Comment by kriswest** [comment]
The missing `id` and `name` properties should all be fixed by : https://github.com/finos/FDC3/pull/1233

**Comment by kriswest** [comment]
Checklist for changes that need to be made to close this (from @hughtroeger very detailed review)

- No Overview page for Context Data Part
  - [ ] Script is currently generating a new section, it should reuse the old section which contains the overview page 
- No Context page for Context Data Part
  - [ ] As discussed, we should just move this into the overview as no one should be broadcasting that base schema. 
- ‘Required’ boolean field is missing for all properties
  - [ ] Needs fixing 
- Schema link goes to github.com instead of fdc3.finos.org, assume that will update on publish/deploy
  - We could/should use whats in the $id field for this rather than the file URL (will need a code change and we don't need to wait for publishing). However, @mistryvinay and I are not sure it matters... Easier to raise a PR to change it with the github URL.
- All reference links are broken in the Netlify preview, is this expected?
  - [ ] Bug: URL is missing the /schemas/ element
- All properties with array types no longer have specific array type (like Instrument[]), they just say ‘array’. Many do have a sentence that says what kind of array it is, and some have reference links to the schema page for the thing it is an array of
  - [ ] Bug: need to add details for the items (which is expressed in the schemas). Some will be base types, others references to other schemas. 
- Properties with a reference link to another docs page no longer have ‘type’ explicitly declared, this is probably fine as the information is present in the page linked
  - [ ] Change the 'Reference' label to 'Type:'
- I find the new format a little difficult to read, I think some indentation for the data under each property would go a long way.
  - [ ] Add indentation to properties

Page specific differences:

Chart

- Instruments property used to have type ‘Instrument[]’ now it has ‘array, is clear from the description that this is an array of instruments
  - Would be fixed by rendering types for arrays
- ChatInitSettings: Message property used to be of type ‘Message’, new one has an example but should probably have a reference link to fdc3.message
  - [ ] Properties defined with `anyOf` and `oneOf` are not being rendered - and probably won't be popular with @bingenito
- Chat Search Criteria: Properties property used to have type (Instrument | Contact | Organization | string)[] , now it just has type ‘array’
  - Will be fixed by types for arrays and support for  `anyOf`/`oneOf` 
- Instrument
  - ‘id’ subproperties used have working links under ‘More Info’, now they have URLs that aren’t active hyperlinks links displayed like <https://www.isin.org/>
    - [ ] Links are being rendered with backticks around them so they appear to be code instead of hrefs - just remove the backticks!
- Order
  - Details.product has type and description ‘undefined’
    - [ ] Sub-properties need to render references to other schemas and skip the title and description if not present as its handled on that page 
- OrderList
  - o ‘orders’ property used to be of type ‘Trade[]’ which I believe is incorrect, I think it should be Order[], now it just has type ‘array’ but does not have a reference link to the order schema page
  - Generation will fix this - it was a mistake in the old docs page (bad Kris)
- TransactionResult
  - o Missing ‘message’ property
    - THis is a straight error in the schema - I'll open a PR to fix this now! Good spot @hughtroeger

**Comment by kriswest** [comment]
@TheJuanAndOnly99 & @finos/fdc3-maintainers  I'm going to merge this into branch on the FINOS repo so we can continue the development there. I'm pretty sure we can get this running with the existing netlify config using npm scripts to trigger it instead (which will be better for local development).

Many thanks for your help with this! We'd happily take more of course ;-) I'll tag you in the new PR once this is moved over.

**Pull request merged**
